gr-digital/lib/decision_feedback_equalizer_impl.cc

```diff
@@ -0,0 +1,267 @@
+/* -*- c++ -*- */
+/*
+ * Copyright 2020 Free Software Foundation, Inc.
+ *
+ * This file is part of GNU Radio
+ *
+ * SPDX-License-Identifier: GPL-3.0-or-later
+ *
+ */
+
+#ifdef HAVE_CONFIG_H
+#include "config.h"
+#endif
+
+#include "adaptive_algorithms.h"
+#include "decision_feedback_equalizer_impl.h"
+#include <gnuradio/io_signature.h>
+#include <volk/volk.h>
+#include <boost/smart_ptr/make_unique.hpp>
+#include <algorithm>
+
+using namespace std;
+namespace gr {
+namespace digital {
+
+decision_feedback_equalizer::sptr
+decision_feedback_equalizer::make(unsigned num_taps_forward,
+                                  unsigned num_taps_feedback,
+                                  unsigned sps,
+                                  adaptive_algorithm_sptr alg,
+                                  bool adapt_after_training,
+                                  std::vector<gr_complex> training_sequence,
+                                  const std::string& training_start_tag)
+{
+    return gnuradio::get_initial_sptr(
+        new decision_feedback_equalizer_impl(num_taps_forward,
+                                             num_taps_feedback,
+                                             sps,
+                                             alg,
+                                             adapt_after_training,
+                                             training_sequence,
```

Comment:
`std::move(training_sequence)`

Removes one copy.
(when a function takes an object that will *always* be copied, then passing it by value (like is already done) is better than by reference.

A smart compiler *may* already see that it can move, here, but adding `std::move` forces it.

You can make the same case for `training_start_tag` being by value instead of constref.

The total reduction of copies if passed by value over constref when callee always makes a copy is when caller calls with an rvalue. Then there will be zero copies in total.

Reply:
I like the suggestion but think I'm going to avoid this for now since it's not a lot of data, and I don't see this pattern (yet) elsewhere in the codebase.